### ISSUE TITLE
KeyRepeatDispatcher related deadlock

### ISSUE THREAD

**Issue by wmww** [MirServer/mir#1617]
I've hit this twice on Arch but I haven't got a reliable reproducer yet. It looks like `device_state.repeat_alarm.reset();` (`key_repeat_dispatcher.cpp:170`) is causing a `TimerGSource` to lock. This is a problem if in another thread the `TimerGSource` is locked and that thread is waiting for the `KeyRepeatDispatcher` to lock. Here's the stack traces of the two relevant threads:
```
Thread 1 "miral-shell.bin" received signal SIGTSTP, Stopped (user).
0x00007f11719f3620 in __lll_lock_wait () from /usr/lib/libpthread.so.0
(gdb) bt
#0  0x00007f11719f3620 in __lll_lock_wait () at /usr/lib/libpthread.so.0
#1  0x00007f11719ebdf3 in pthread_mutex_lock () at /usr/lib/libpthread.so.0
#2  0x00007f1170d60843 in __gthread_mutex_lock(__gthread_mutex_t*) (__mutex=0x564bbb0f08e8) at /usr/include/c++/10.1.0/x86_64-pc-linux-gnu/bits/gthr-default.h:749
#3  0x00007f1170d60894 in std::mutex::lock() (this=0x564bbb0f08e8) at /usr/include/c++/10.1.0/bits/std_mutex.h:100
#4  0x00007f1170d608f0 in std::lock_guard<std::mutex>::lock_guard(std::mutex&) (this=0x7ffda8993e30, __m=...) at /usr/include/c++/10.1.0/bits/std_mutex.h:159
#5  0x00007f1170e132d0 in operator()() const (__closure=0x7f113c023790) at /home/wmww/code/mir/src/server/input/key_repeat_dispatcher.cpp:208
#6  0x00007f1170e13ce4 in std::__invoke_impl<void, mir::input::KeyRepeatDispatcher::handle_key_input(MirInputDeviceId, const MirKeyboardEvent*)::<lambda()>&>(std::__invoke_other, struct {...} &) (__f=...)
    at /usr/include/c++/10.1.0/bits/invoke.h:60
#7  0x00007f1170e13b7b in std::__invoke_r<void, mir::input::KeyRepeatDispatcher::handle_key_input(MirInputDeviceId, const MirKeyboardEvent*)::<lambda()>&>(struct {...} &) (__fn=...)
    at /usr/include/c++/10.1.0/bits/invoke.h:153
#8  0x00007f1170e139ed in std::_Function_handler<void(), mir::input::KeyRepeatDispatcher::handle_key_input(MirInputDeviceId, const MirKeyboardEvent*)::<lambda()> >::_M_invoke(const std::_Any_data &)
    (__functor=...) at /usr/include/c++/10.1.0/bits/std_function.h:291
#9  0x00007f1170d6308c in std::function<void ()>::operator()() const (this=0x7f113c008988) at /usr/include/c++/10.1.0/bits/std_function.h:622
#10 0x00007f1170dc3f6a in mir::BasicCallback::operator()() (this=0x7f113c008980) at /home/wmww/code/mir/src/server/basic_callback.cpp:28
#11 0x00007f1170dc3947 in mir::LockableCallbackWrapper::operator()() (this=0x7f113c02e120) at /home/wmww/code/mir/src/server/lockable_callback_wrapper.cpp:41
#12 0x00007f1170d888f7 in mir::detail::TimerGSource::dispatch(GSource*, GSourceFunc, gpointer) (source=0x564bbb2be2f0) at /home/wmww/code/mir/src/server/glib_main_loop_sources.cpp:322
#13 0x00007f11717ed43c in g_main_context_dispatch () at /usr/lib/libglib-2.0.so.0
#14 0x00007f117183b1d9 in  () at /usr/lib/libglib-2.0.so.0
#15 0x00007f11717ec221 in g_main_context_iteration () at /usr/lib/libglib-2.0.so.0
#16 0x00007f1170d7a68f in mir::GLibMainLoop::run() (this=0x564bbaea5f90) at /home/wmww/code/mir/src/server/glib_main_loop.cpp:146
#17 0x00007f1170d60b0c in mir::DisplayServer::run() (this=0x7ffda8994138) at /home/wmww/code/mir/src/server/display_server.cpp:237
#18 0x00007f1170d5b790 in mir::run_mir(mir::ServerConfiguration&, std::function<void (mir::DisplayServer&)>, std::function<void (int)> const&) (config=..., init=..., terminator=...)
    at /home/wmww/code/mir/src/server/run_mir.cpp:110
#19 0x00007f1170d959c0 in mir::Server::run() (this=0x564bbae91fd0) at /home/wmww/code/mir/src/server/server.cpp:412
#20 0x00007f1171a8ba71 in miral::MirRunner::Self::run_with(std::initializer_list<std::function<void (mir::Server&)> >) (this=0x564bbae8ce10, options=...) at /home/wmww/code/mir/src/miral/runner.cpp:153
#21 0x00007f1171a8bfc9 in miral::MirRunner::run_with(std::initializer_list<std::function<void (mir::Server&)> >) (this=0x7ffda8994380, options=...) at /home/wmww/code/mir/src/miral/runner.cpp:200
#22 0x0000564bb90e2457 in main(int, char const**) (argc=1, argv=0x7ffda89949e8) at /home/wmww/code/mir/examples/miral-shell/shell_main.cpp:114
(gdb) info threads
  Id   Target Id                                             Frame 
* 1    Thread 0x7f116fa6a880 (LWP 1124188) "miral-shell.bin" 0x00007f11719f3620 in __lll_lock_wait () from /usr/lib/libpthread.so.0
  2    Thread 0x7f116fa66700 (LWP 1124210) "miral-shell.bin" 0x00007f11714c471d in syscall () from /usr/lib/libc.so.6
  3    Thread 0x7f116f265700 (LWP 1124211) "miral-shell.bin" 0x00007f11714c471d in syscall () from /usr/lib/libc.so.6
  4    Thread 0x7f1163669700 (LWP 1124213) "miral-s:disk$0"  0x00007f11719efe32 in pthread_cond_wait@@GLIBC_2.3.2 () from /usr/lib/libpthread.so.0
  5    Thread 0x7f1162e68700 (LWP 1124214) "miral-s:disk$1"  0x00007f11719efe32 in pthread_cond_wait@@GLIBC_2.3.2 () from /usr/lib/libpthread.so.0
  6    Thread 0x7f1162667700 (LWP 1124215) "miral-s:disk$2"  0x00007f11719efe32 in pthread_cond_wait@@GLIBC_2.3.2 () from /usr/lib/libpthread.so.0
  7    Thread 0x7f1161e66700 (LWP 1124216) "miral-s:disk$3"  0x00007f11719efe32 in pthread_cond_wait@@GLIBC_2.3.2 () from /usr/lib/libpthread.so.0
  8    Thread 0x7f116136c700 (LWP 1124217) "miral-shell.bin" 0x00007f11719efe32 in pthread_cond_wait@@GLIBC_2.3.2 () from /usr/lib/libpthread.so.0
  9    Thread 0x7f1160a88700 (LWP 1124218) "Mir/Snapshot"    0x00007f11719efe32 in pthread_cond_wait@@GLIBC_2.3.2 () from /usr/lib/libpthread.so.0
  10   Thread 0x7f114ffff700 (LWP 1124219) "Mir/Comp"        0x00007f11719efe32 in pthread_cond_wait@@GLIBC_2.3.2 () from /usr/lib/libpthread.so.0
  11   Thread 0x7f114f2fe700 (LWP 1124220) "Mir/Input Reade" 0x00007f11719f3620 in __lll_lock_wait () from /usr/lib/libpthread.so.0
  12   Thread 0x7f114eafd700 (LWP 1124221) "Mir/Wayland"     0x00007f11714c9f3e in epoll_wait () from /usr/lib/libc.so.6
  13   Thread 0x7f114e2fc700 (LWP 1124222) "Mir/X11 Spawner" 0x00007f11714bf05f in poll () from /usr/lib/libc.so.6
  14   Thread 0x7f114dafb700 (LWP 1124223) "miral-shell.bin" 0x00007f11714bf05f in poll () from /usr/lib/libc.so.6
  15   Thread 0x7f11319e6700 (LWP 1124461) "Mir/X11 WM Read" 0x00007f11714bf05f in poll () from /usr/lib/libc.so.6
(gdb) thread 11
[Switching to thread 11 (Thread 0x7f114f2fe700 (LWP 1124220))]
#0  0x00007f11719f3620 in __lll_lock_wait () from /usr/lib/libpthread.so.0
(gdb) bt
#0  0x00007f11719f3620 in __lll_lock_wait () at /usr/lib/libpthread.so.0
#1  0x00007f11719ebe81 in pthread_mutex_lock () at /usr/lib/libpthread.so.0
#2  0x00007f1170d87ed1 in __gthread_mutex_lock(__gthread_mutex_t*) (__mutex=0x564bbb2be398) at /usr/include/c++/10.1.0/x86_64-pc-linux-gnu/bits/gthr-default.h:749
#3  0x00007f1170d8b5e8 in __gthread_recursive_mutex_lock(__gthread_recursive_mutex_t*) (__mutex=0x564bbb2be398) at /usr/include/c++/10.1.0/x86_64-pc-linux-gnu/bits/gthr-default.h:811
#4  0x00007f1170d8bb66 in std::recursive_mutex::lock() (this=0x564bbb2be398) at /usr/include/c++/10.1.0/mutex:106
#5  0x00007f1170d8ca92 in std::lock_guard<std::recursive_mutex>::lock_guard(std::recursive_mutex&) (this=0x7f114f2fd2b8, __m=...) at /usr/include/c++/10.1.0/bits/std_mutex.h:159
#6  0x00007f1170d88a37 in mir::detail::TimerGSource::disable(GSource*) (source=0x564bbb2be2f0) at /home/wmww/code/mir/src/server/glib_main_loop_sources.cpp:342
#7  0x00007f1170d88a7e in operator()(GSource*) const (__closure=0x7f113c03a488, gsource=0x564bbb2be2f0) at /home/wmww/code/mir/src/server/glib_main_loop_sources.cpp:358
#8  0x00007f1170d8af8a in std::__invoke_impl<void, mir::detail::add_timer_gsource(GMainContext*, const std::shared_ptr<mir::time::Clock>&, const std::shared_ptr<mir::LockableCallback>&, const std::function<void()>&, mir::time::Timestamp)::<lambda(GSource*)>&, _GSource*>(std::__invoke_other, struct {...} &) (__f=...) at /usr/include/c++/10.1.0/bits/invoke.h:60
#9  0x00007f1170d8a76a in std::__invoke_r<void, mir::detail::add_timer_gsource(GMainContext*, const std::shared_ptr<mir::time::Clock>&, const std::shared_ptr<mir::LockableCallback>&, const std::function<void()>&, mir::time::Timestamp)::<lambda(GSource*)>&, _GSource*>(struct {...} &) (__fn=...) at /usr/include/c++/10.1.0/bits/invoke.h:153
#10 0x00007f1170d89f7f in std::_Function_handler<void(_GSource*), mir::detail::add_timer_gsource(GMainContext*, const std::shared_ptr<mir::time::Clock>&, const std::shared_ptr<mir::LockableCallback>&, const std::function<void()>&, mir::time::Timestamp)::<lambda(GSource*)> >::_M_invoke(const std::_Any_data &, _GSource *&&) (__functor=..., __args#0=@0x7f114f2fd390: 0x564bbb2be2f0)
    at /usr/include/c++/10.1.0/bits/std_function.h:291
#11 0x00007f1170d8c8bf in std::function<void (_GSource*)>::operator()(_GSource*) const (this=0x7f113c03a488, __args#0=0x564bbb2be2f0) at /usr/include/c++/10.1.0/bits/std_function.h:622
#12 0x00007f1170d8825c in mir::detail::GSourceHandle::ensure_no_further_dispatch() (this=0x7f113c03a480) at /home/wmww/code/mir/src/server/glib_main_loop_sources.cpp:143
#13 0x00007f1170d79f9c in (anonymous namespace)::AlarmImpl::~AlarmImpl() (this=0x7f113c03a400, __in_chrg=<optimized out>) at /home/wmww/code/mir/src/server/glib_main_loop.cpp:53
#14 0x00007f1170d7a014 in (anonymous namespace)::AlarmImpl::~AlarmImpl() (this=0x7f113c03a400, __in_chrg=<optimized out>) at /home/wmww/code/mir/src/server/glib_main_loop.cpp:55
#15 0x00007f1170e15286 in std::default_delete<mir::time::Alarm>::operator()(mir::time::Alarm*) const (this=0x7f113c015ac0, __ptr=0x7f113c03a400) at /usr/include/c++/10.1.0/bits/unique_ptr.h:84
#16 0x00007f1170e17bc4 in std::_Sp_counted_deleter<mir::time::Alarm*, std::default_delete<mir::time::Alarm>, std::allocator<void>, (__gnu_cxx::_Lock_policy)2>::_M_dispose() (this=0x7f113c015ab0)
    at /usr/include/c++/10.1.0/bits/shared_ptr_base.h:474
#17 0x00007f1170d5c45f in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release() (this=0x7f113c015ab0) at /usr/include/c++/10.1.0/bits/shared_ptr_base.h:158
#18 0x00007f1170d5c0b1 in std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count() (this=0x7f114f2fd528, __in_chrg=<optimized out>) at /usr/include/c++/10.1.0/bits/shared_ptr_base.h:733
#19 0x00007f1170e147f2 in std::__shared_ptr<mir::time::Alarm, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr() (this=0x7f114f2fd520, __in_chrg=<optimized out>) at /usr/include/c++/10.1.0/bits/shared_ptr_base.h:1183
#20 0x00007f1170e14b82 in std::__shared_ptr<mir::time::Alarm, (__gnu_cxx::_Lock_policy)2>::reset() (this=0x7f113c015450) at /usr/include/c++/10.1.0/bits/shared_ptr_base.h:1301
#21 0x00007f1170e1341d in mir::input::KeyRepeatDispatcher::handle_key_input(long, MirKeyboardEvent const*) (this=0x564bbb0f08e0, id=1, kev=0x7f113c028be0)
    at /home/wmww/code/mir/src/server/input/key_repeat_dispatcher.cpp:170
#22 0x00007f1170e13069 in mir::input::KeyRepeatDispatcher::dispatch(std::shared_ptr<MirEvent const> const&) (this=0x564bbb0f08e0, event=std::shared_ptr<const MirEvent> (use count 2, weak count 0) = {...})
    at /home/wmww/code/mir/src/server/input/key_repeat_dispatcher.cpp:150
#23 0x00007f1170e18758 in mir::input::SeatInputDeviceTracker::dispatch(std::shared_ptr<MirEvent> const&) (this=0x564bbaebce18, event=std::shared_ptr<MirEvent> (use count 2, weak count 0) = {...})
    at /home/wmww/code/mir/src/server/input/seat_input_device_tracker.cpp:110
#24 0x00007f1170dc4227 in mir::input::BasicSeat::dispatch_event(std::shared_ptr<MirEvent> const&) (this=0x564bbaebce10, event=std::shared_ptr<MirEvent> (use count 2, weak count 0) = {...})
    at /home/wmww/code/mir/src/server/input/basic_seat.cpp:204
#25 0x00007f1170df72b0 in mir::input::DefaultInputDeviceHub::RegisteredDevice::handle_input(std::shared_ptr<MirEvent> const&)
    (this=0x7f113c001d50, event=std::shared_ptr<MirEvent> (use count 2, weak count 0) = {...}) at /home/wmww/code/mir/src/server/input/default_input_device_hub.cpp:291
#26 0x00007f116e76e75d in mir::input::X::XInputDevice::key_release(std::chrono::duration<long, std::ratio<1l, 1000000000l> >, unsigned int, int) (this=0x564bbb2bb0e0, event_time=..., key_sym=65288, key_code=14)
    at /home/wmww/code/mir/src/platforms/x11/input/input_device.cpp:156
#27 0x00007f116e76ba34 in mir::input::X::XInputPlatform::process_input_event() (this=0x564bbb2bb040) at /home/wmww/code/mir/src/platforms/x11/input/input_platform.cpp:211
#28 0x00007f116e76b065 in operator()() const (__closure=0x564bbb2b8d78) at /home/wmww/code/mir/src/platforms/x11/input/input_platform.cpp:76
#29 0x00007f116e76c85c in std::__invoke_impl<void, mir::input::X::XInputPlatform::XInputPlatform(const std::shared_ptr<mir::input::InputDeviceRegistry>&, const std::shared_ptr<_XDisplay>&)::<lambda()>&>(std::__invoke_other, struct {...} &) (__f=...) at /usr/include/c++/10.1.0/bits/invoke.h:60
#30 0x00007f116e76c74d in std::__invoke_r<void, mir::input::X::XInputPlatform::XInputPlatform(const std::shared_ptr<mir::input::InputDeviceRegistry>&, const std::shared_ptr<_XDisplay>&)::<lambda()>&>(struct {...} &) (__fn=...) at /usr/include/c++/10.1.0/bits/invoke.h:153
#31 0x00007f116e76c63e in std::_Function_handler<void(), mir::input::X::XInputPlatform::XInputPlatform(const std::shared_ptr<mir::input::InputDeviceRegistry>&, const std::shared_ptr<_XDisplay>&)::<lambda()> >::_M_invoke(const std::_Any_data &) (__functor=...) at /usr/include/c++/10.1.0/bits/std_function.h:291
#32 0x00007f1170a1770e in std::function<void ()>::operator()() const (this=0x564bbb2b8d78) at /usr/include/c++/10.1.0/bits/std_function.h:622
#33 0x00007f1170a1c0f0 in mir::dispatch::ReadableFd::dispatch(unsigned int) (this=0x564bbb2b8d60, events=1) at /home/wmww/code/mir/src/common/dispatch/readable_fd.cpp:38
#34 0x00007f1170a192f8 in mir::dispatch::MultiplexingDispatchable::dispatch(unsigned int) (this=0x564bbb2b9640, events=1) at /home/wmww/code/mir/src/common/dispatch/multiplexing_dispatchable.cpp:140
#35 0x00007f1170a192f8 in mir::dispatch::MultiplexingDispatchable::dispatch(unsigned int) (this=0x564bbb2bc9a0, events=1) at /home/wmww/code/mir/src/common/dispatch/multiplexing_dispatchable.cpp:140
#36 0x00007f1170a1c3e2 in (anonymous namespace)::dispatch_loop(const std::string &, std::shared_ptr<mir::dispatch::ThreadedDispatcher::ThreadShutdownRequestHandler>, std::shared_ptr<mir::dispatch::Dispatchable>,--Type <RET> for more, q to quit, c to continue without paging--
 const std::function<void()> &) (name="Mir/Input Reader", thread_register=std::shared_ptr<mir::dispatch::ThreadedDispatcher::ThreadShutdownRequestHandler> (use count 4, weak count 0) = {...}, dispatcher=
    std::shared_ptr<mir::dispatch::Dispatchable> (use count 2, weak count 0) = {...}, exception_handler=...) at /home/wmww/code/mir/src/common/dispatch/threaded_dispatcher.cpp:211
#37 0x00007f1170a27bd0 in std::__invoke_impl<void, void (*)(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::shared_ptr<mir::dispatch::ThreadedDispatcher::ThreadShutdownRequestHandler>, std::shared_ptr<mir::dispatch::Dispatchable>, std::function<void ()> const&), std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::shared_ptr<mir::dispatch::ThreadedDispatcher::ThreadShutdownRequestHandler>, std::shared_ptr<mir::dispatch::MultiplexingDispatchable>, std::function<void ()> >(std::__invoke_other, void (*&&)(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::shared_ptr<mir::dispatch::ThreadedDispatcher::ThreadShutdownRequestHandler>, std::shared_ptr<mir::dispatch::Dispatchable>, std::function<void ()> const&), std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&&, std::shared_ptr<mir::dispatch::ThreadedDispatcher::ThreadShutdownRequestHandler>&&, std::shared_ptr<mir::dispatch::MultiplexingDispatchable>&&, std::function<void ()>&&) (__f=
    @0x564bbb2bcb08: 0x7f1170a1c235 <(anonymous namespace)::dispatch_loop(const std::string &, std::shared_ptr<mir::dispatch::ThreadedDispatcher::ThreadShutdownRequestHandler>, std::shared_ptr<mir::dispatch::Dispatchable>, const std::function<void()> &)>) at /usr/include/c++/10.1.0/bits/invoke.h:60
#38 0x00007f1170a27a1b in std::__invoke<void (*)(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::shared_ptr<mir::dispatch::ThreadedDispatcher::ThreadShutdownRequestHandler>, std::shared_ptr<mir::dispatch::Dispatchable>, std::function<void ()> const&), std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::shared_ptr<mir::dispatch::ThreadedDispatcher::ThreadShutdownRequestHandler>, std::shared_ptr<mir::dispatch::MultiplexingDispatchable>, std::function<void ()> >(void (*&&)(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::shared_ptr<mir::dispatch::ThreadedDispatcher::ThreadShutdownRequestHandler>, std::shared_ptr<mir::dispatch::Dispatchable>, std::function<void ()> const&), std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&&, std::shared_ptr<mir::dispatch::ThreadedDispatcher::ThreadShutdownRequestHandler>&&, std::shared_ptr<mir::dispatch::MultiplexingDispatchable>&&, std::function<void ()>&&)
    (__fn=
    @0x564bbb2bcb08: 0x7f1170a1c235 <(anonymous namespace)::dispatch_loop(const std::string &, std::shared_ptr<mir::dispatch::ThreadedDispatcher::ThreadShutdownRequestHandler>, std::shared_ptr<mir::dispatch::Dispatchable>, const std::function<void()> &)>) at /usr/include/c++/10.1.0/bits/invoke.h:95
#39 0x00007f1170a27873 in std::thread::_Invoker<std::tuple<void (*)(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::shared_ptr<mir::dispatch::ThreadedDispatcher::ThreadShutdownRequestHandler>, std::shared_ptr<mir::dispatch::Dispatchable>, std::function<void ()> const&), std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::shared_ptr<mir::dispatch::ThreadedDispatcher::ThreadShutdownRequestHandler>, std::shared_ptr<mir::dispatch::MultiplexingDispatchable>, std::function<void ()> > >::_M_invoke<0ul, 1ul, 2ul, 3ul, 4ul>(std::_Index_tuple<0ul, 1ul, 2ul, 3ul, 4ul>) (this=0x564bbb2bcaa8) at /usr/include/c++/10.1.0/thread:264
#40 0x00007f1170a2751c in std::thread::_Invoker<std::tuple<void (*)(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::shared_ptr<mir::dispatch::ThreadedDispatcher::ThreadShutdownRequestHandler>, std::shared_ptr<mir::dispatch::Dispatchable>, std::function<void ()> const&), std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::shared_ptr<mir::dispatch::ThreadedDispatcher::ThreadShutdownRequestHandler>, std::shared_ptr<mir::dispatch::MultiplexingDispatchable>, std::function<void ()> > >::operator()() (this=0x564bbb2bcaa8)
    at /usr/include/c++/10.1.0/thread:271
#41 0x00007f1170a26e8e in std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::shared_ptr<mir::dispatch::ThreadedDispatcher::ThreadShutdownRequestHandler>, std::shared_ptr<mir::dispatch::Dispatchable>, std::function<void ()> const&), std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::shared_ptr<mir::dispatch::ThreadedDispatcher::ThreadShutdownRequestHandler>, std::shared_ptr<mir::dispatch::MultiplexingDispatchable>, std::function<void ()> > > >::_M_run() (this=0x564bbb2bcaa0)
    at /usr/include/c++/10.1.0/thread:215
#42 0x00007f117167ab74 in std::execute_native_thread_routine(void*) (__p=0x564bbb2bcaa0) at /build/gcc/src/gcc/libstdc++-v3/src/c++11/thread.cc:80
#43 0x00007f11719e9422 in start_thread () at /usr/lib/libpthread.so.0
#44 0x00007f11714c9bf3 in clone () at /usr/lib/libc.so.6
```

**Comment by wmww** [comment]
This should probably block #1613

**Comment by AlanGriffiths** [comment]
> This should probably block #1613

Agreed. We should at least spend some time to understand this before release.

**Comment by wmww** [comment]
I've reproduced it a couple more times including on Ubuntu. The issue occurs when a repeat timer fires at the same time as a key-up (which cancels the timer). To reproduce, an app, hold a key long enough for it to start repeating and then let go. Keep doing that until the timing lines up and Mir stops accepting input. I've found that generally the repeats keep firing and the rest of Mir keeps working, so you see letters keep appearing like the key's stuck down.
